test/e2e/password_grant_check_test.go

```diff
@@ -32,3 +38,82 @@ func TestGitLabAsOIDCPasswordGrantCheck(t *testing.T) {
 
 	require.Equal(t, true, gitlabIDPConfig.UseAsChallenger, "gitlab is not configured as challenger")
 }
+
+func TestKeycloakAsOIDCPasswordGrantCheck(t *testing.T) {
+	testContext := context.Background()
+
+	kubeConfig := test.NewClientConfigForTest(t)
+
+	kubeClients, err := kubernetes.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	configClient, err := configv1client.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	_, idpName, cleanups := test.AddKeycloakIDP(t, kubeConfig)
+	defer library.IDPCleanupWrapper(func() {
+		for _, c := range cleanups {
+			c()
+		}
+	})()
+
+	config, err := library.GrabOAuthServerConfig(kubeClients.CoreV1())
+	require.NoError(t, err)
+
+	kcIDPConfig := library.GetIDPByName(config, idpName)
+	require.NotNil(t, kcIDPConfig, "did not find idp %q in the config: %#v", idpName, config)
+
+	require.Equal(t, false, kcIDPConfig.UseAsChallenger, "keycloak is configured as challenger but it should not be")
+
+	oauthConfig, err := configClient.OAuths().Get(testContext, "cluster", v1.GetOptions{})
+	require.NoError(t, err)
+
+	var configIDP configv1.IdentityProvider
+	for _, idp := range oauthConfig.Spec.IdentityProviders {
+		if idp.Name == idpName {
+			configIDP = idp
+			break
+		}
+	}
+
+	// Get a keycloak client for the external KC URL
+	// skip TLS verify on the route's cert
+	insecureConfig := rest.CopyConfig(kubeConfig)
+	insecureConfig.Insecure = true
+	insecureTransport, err := rest.TransportFor(insecureConfig)
+	require.NoError(t, err)
+	// finally get the client
+	kcClient := library.KeycloakClientFor(t, insecureTransport, configIDP.OpenID.Issuer, "master")
+	err = kcClient.AuthenticatePassword("admin-cli", "", "admin", "password")
+	require.NoError(t, err)
+
+	client, err := kcClient.GetClientByClientID(configIDP.OpenID.ClientID)
+	require.NoError(t, err)
+
+	err = kcClient.UpdateClientDirectAccessGrantsEnabled(client["id"].(string), true)
+	require.NoError(t, err)
+
+	// bump the configured secret since the operator caches the password grant check
+	// based on it
+	configSecretName := configIDP.OpenID.ClientSecret.Name
+	configSecret, err := kubeClients.CoreV1().Secrets("openshift-config").Get(testContext, configSecretName, v1.GetOptions{})
+	require.NoError(t, err)
+
+	if configSecret.Annotations == nil {
+		configSecret.Annotations = map[string]string{}
+	}
+	configSecret.Annotations["bumped"] = "true"
+	_, err = kubeClients.CoreV1().Secrets("openshift-config").Update(testContext, configSecret, v1.UpdateOptions{})
+	require.NoError(t, err)
+
+	err = library.WaitForOperatorToPickUpChanges(t, configClient, "authentication")
```

Comment:
is this race free to wait after the change above just a moment ago?

Reply:
I don't think it's exactly race free, but until now I think it has always worked.

Reply:
but it seems that perhaps this time, it fails in CI, yet it never failed for me. I'll try to look for a better way.

Reply:
actually, I can see that the operator went degraded, but this should otherwise still be safe. I added degraded reporting for this and I'll check why that is.

Reply:
Found the issue, I forgot to wire gitlab cleanup with the new refactor. This method of checking whether the operator picked up the changes still seems safe, then.